packages/gguf/src/gguf.ts

```diff
@@ -0,0 +1,275 @@
+type MetadataBaseValue = string | number | bigint | boolean;
+type MetadataValue = MetadataBaseValue | MetadataBaseValue[] | MetadataValue[]; /// recursive as arrays can be nested.
+
+type Version = 1 | 2 | 3;
+const isVersion = (version: number): version is Version => version === 1 || version === 2 || version === 3;
+
+const ggufMagicNumber = new Uint8Array([0x47, 0x47, 0x55, 0x46]); /// "GGUF"
+
+export enum GGMLQuantizationType {
+	F32 = 0,
+	F16 = 1,
+	Q4_0 = 2,
+	Q4_1 = 3,
+	Q5_0 = 6,
+	Q5_1 = 7,
+	Q8_0 = 8,
+	Q8_1 = 9,
+	Q2_K = 10,
+	Q3_K = 11,
+	Q4_K = 12,
+	Q5_K = 13,
+	Q6_K = 14,
+	Q8_K = 15,
+	IQ2_XXS = 16,
+	IQ2_XS = 17,
+	IQ3_XXS = 18,
+	IQ1_S = 19,
+	IQ4_NL = 20,
+	IQ3_S = 21,
+	IQ2_S = 22,
+	IQ4_XS = 23,
+}
+
+enum GGUFValueType {
+	UINT8 = 0,
+	INT8 = 1,
+	UINT16 = 2,
+	INT16 = 3,
+	UINT32 = 4,
+	INT32 = 5,
+	FLOAT32 = 6,
+	BOOL = 7,
+	STRING = 8,
+	ARRAY = 9,
+	UINT64 = 10,
+	INT64 = 11,
+	FLOAT64 = 12,
+}
+function isGGUFValueType(n: number): n is GGUFValueType {
+	return typeof GGUFValueType[n] === "string";
+}
+
+const HTTP_CHUNK_SIZE = 2 * 10 ** 6; /// 2MB
+const HTTP_DATA_LEEWAY = 5 * 10 ** 5; /// 500kb
+
+/**
+ * Internal stateful instance to fetch ranges of HTTP data when needed
+ */
+class RangeView {
+	private chunk: number;
+	private buffer: ArrayBuffer;
+
+	readonly view: DataView;
+
+	constructor(public url: string) {
```

Comment:
```suggestion
	constructor(public url: string, private headers?: Record<string, string>) {
```

regarding [this comment](https://github.com/huggingface/huggingface.js/pull/540#issuecomment-1986870330), `RangeView` can have private property `headers` where Authorization header, amongst other header options, could be passed. Afterwards, `this.headers` could be used inside `fetch` inside `fetchChunk`

Reply:
i'd rather we pass an optional `fetch` like we do for safetensors:

https://github.com/huggingface/huggingface.js/blob/c26f6e0d9d26b0c0fe68ee8622d3a32433650d80/packages/hub/src/lib/parse-safetensors-metadata.ts#L63-L66

WDYT @coyotte508 @mishig25?

Reply:
will be more future-proof/flexible

Reply:
can you possibly make this change if you agree @mishig25?

Reply:
yes, agree

Reply:
https://github.com/huggingface/huggingface.js/pull/543